cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
+	// Nominal version of Kubernetes control-plane to run.
+	KubernetesVersion KubernetesVersionInfo
```

Comment:
Does this control-plane version apply strictly to APIserver/controller-manager/scheduler or it is expected to have some impacts also on the version of other other "core" components (kube-dns, kube-proxy, Kubelet, Etcd)?
Somehow related to his, IMO the comment in this API should also state what kind of version-skew policy is expected to be enforced by controllers

fyi in kubadm now 
- kube-dns version is linked to controlplane version
- kube-proxy version is the same of control plane version
- etcd version is a constant (and thus linked to kubeadm version)
- kubelet is not managed

Reply:
The current thought is that kubelet is controlled by the version info in the machines API. For the time being, the provider should pick appropriate versions of kube-dns and kube-proxy to satisfy this version. We may have to revise that when/if we progress with an addon management solution. ETCD is technically and implementation detail that I would not like to expose through this API. So the provider should pick what it wants.

That's not to say a provider couldn't let users specify such things in the provider config field. If it is common and universally applicable enough, we can add it to the main API later. It's easier to add than to take away.